### ISSUE TITLE
Message handlers can fail but puback already sent to server

### ISSUE THREAD

**Issue by jdiamond** [mqttjs/MQTT.js#209]
At qos 1+, shouldn't the message handler have a chance to succeed or fail before acknowledging the message was successfully received?

**Comment by mcollina** [comment]
Released in 1.0.0. See https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L613-L623.

**Comment by rossjourdain** [comment]
Hey @mcollina, I'd quite like to use this but I'm not sure how to prevent the puback from going out if I haven't managed to take care of the message yet.  Is that possible now?

**Comment by mcollina** [comment]
@rossjourdain it has been possible for over a year. Maybe the docs are not great (please send PR :P), and definitely there is an issue with #425. Add an `handleMessage` function to the client, and call the callback when you are done. 

Any contribution/help into the project is welcomed!

**Comment by rossjourdain** [comment]
I'm more that happy to contribute how ever I can once I get my head around this ;)
I can't seem to find any examples of the handleMessage process.

When I set the handleMessage function, it fires no trouble but the puback packet goes out regardless.  From my reading of the code, there's no way anything but that can happen.

I guess one way to do it is to make handleMessage take a success callback as well as a failure callback.  Something like this?

```
client.handleMessage = function(packet, success_cb, failure_cb) {
  console.log('handleMessage callback: ' + ' ' + packet.cmd);

  if(processTheMessage) {
    success_cb();
  } else {
    failure_cb():
  }
}
```

While the client.js side is something like this:

Modified version of:
https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L776-L800

```
switch (qos) {
    case 2:
      this.incomingStore.put(packet, function () {
        that._sendPacket({cmd: 'pubrec', messageId: mid}, done);
      });
      break;
    case 1:
      this.emit('message', topic, message, packet);
      // Send the puback on success only
      this.handleMessage(packet, function(){
        that._sendPacket({cmd: 'puback', messageId: mid}, done);
      }, done);
      break;
    case 0:
      this.emit('message', topic, message, packet);
      this.handleMessage(packet, done, done);
      break;
    default:
      // do nothing
      // log or throw an error about unknown qos
      break;
  }
```

Does that make any sense at all?

**Comment by mcollina** [comment]
Just to clarify, I think that's not ok to have two functions.
I think that we **must** ACK a message, independently of the error. If the callback is called with an `'error'`, we should probably forward that `'error'` to the client itself.

**Comment by mcollina** [comment]
Fixed in #697
